src/ucp/tag/rndv.c

```diff
@@ -596,6 +603,236 @@ static void ucp_rndv_req_send_rma_get(ucp_request_t *rndv_req, ucp_request_t *rr
     ucp_request_send(rndv_req, 0);
 }
 
+UCS_PROFILE_FUNC_VOID(ucp_rndv_recv_frag_put_completion, (self, status),
+                      uct_completion_t *self, ucs_status_t status)
+{
+    ucp_request_t *freq     = ucs_container_of(self, ucp_request_t,
+                                               send.state.uct_comp);
+    ucp_request_t *req      = freq->send.rndv_put.sreq;
+    ucp_request_t *rndv_req = (ucp_request_t*)freq->send.rndv_put.remote_request;
+
+    ucs_trace_req("freq:%p: recv_frag_put done. rreq:%p ", freq, req);
+
+    /* release memory descriptor */
+    if (freq->send.mdesc) {
+        ucs_mpool_put_inline((void *)freq->send.mdesc);
+    }
+
+    if (rndv_req) {
+        /* pipeline recv get protocol */
+        rndv_req->send.state.dt.offset += freq->send.length;
+
+        /* send ATS for fragment get rndv completion */
+        if (rndv_req->send.length == rndv_req->send.state.dt.offset) {
+            ucp_rkey_destroy(rndv_req->send.rndv_get.rkey);
+            ucp_rndv_req_send_ats(rndv_req, req,
+                                  rndv_req->send.rndv_get.remote_request, UCS_OK);
```

Comment:
`rndv_req->send.rndv_put.remote_request`

Reply:
here, put is completed in frreq. rndv_req is saved request object which is handling original get request from remote.  `rndv_req->send.rndv_get.remote_request` is valid here which is set at line:775